City-of-Helsinki/parking-permits
# PV-402: Add support for searching orders

## Description

Adds support for searching orders by the parameters mentioned in [PV-402](https://helsinkisolutionoffice.atlassian.net/browse/PV-402). They are:
- parking permit validity date range
- parking zone
- text search
- contract type
- payment type
- parking discounts (low emission vehicles).

This also adds a low emission field vehicles, which gets updated every time you save the model, and makes the customer's national ID number searchable, which requires an irreversible data migration.

## Context

In order for the admin UI to be able to search the orders by the given parameters, the back-end needs to support this.

[PV-402](https://helsinkisolutionoffice.atlassian.net/browse/PV-402)

## How Has This Been Tested?

There are unit tests for both the low emission and the data range, since their logic isn't exactly straight-forward. For the rest of the search parameters, I'll admit I followed suit and didn't include any unit tests; they're pretty simple.

For migrations, I tested them manually. Since we don'´t have any real production data yet, I don't think adding a very slow migration test would bring much value.

## Manual Testing Instructions for Reviewers

I'd say test it through the admin UI, but since that side isn't ready yet, you'll need to test it with your own queries.

Before starting the queries, run the migrations and check from Django admin that the customers' national ID numbers look fine (i.e. they aren't corrupted) and that the vehicles seem to have the correct low emission status. Probably no need to go through them all, but if there's a non-low emission electric vehicle, something probably went wrong with the migrations.

With order search, test the following:
- `q` filters by first name, last name, vehicle registration number and national ID number
- `startDate`/`endDate` doesn't crash anything (the logic is covered by unit tests)
- `parkingZone` filters by parking zone name
- `contractTypes` filters by contract type (`OPEN_ENDED`, `FIXED_PERIOD`)
- `paymentTypes` filters by payment type (`ONLINE_PAYMENT`, `CASHIER_PAYMENT`)
- `pricedDiscounts` filters by price discount (`LOW_EMISSION`)

With the low emission field, test this in Django admin:
- Create a new vehicle, make it electric, save
- Confirm that the new vehicle is low emission
- Change the vehicle to fulfill high emission vehicle criteria (changing from electric to something and omitting emission is enough) and save
- Confirm that the vehicle is not low emission any more.


## Screenshots

No screenshots.
